### PR TITLE
Adds a fuzz test for test_precision

### LINKED ISSUE #14691: Integrating with Google/OSS-fuzz
### What is the problem this feature will solve?

I would like to help keep astropy running smoothly by adding fuzz testing and integrating it with Google/OSS-fuzz. In case you are not familiar with fuzz testing, briefly, it is an additional layer of testing (working alongside other tests) that is particularly good at finding unexpected bugs as it doesn't rely on a developer's suspicion to target a piece of code for testing. [Wikipedia ](https://en.wikipedia.org/wiki/Fuzzing)explains it nicely:

>In programming and software development, fuzzing or fuzz testing is an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program. The program is then monitored for exceptions such as crashes, failing built-in code assertions, or potential memory leaks.

### Describe the desired outcome

Google offers a free, continuous fuzzing service called [OSS-fuzz.](https://github.com/google/oss-fuzz) If astropy is integrated into oss-fuzz, the fuzz tests under astropy will be built and then run once a day, to search for bugs and vulnerabilities. This service can be integrated with the CI for astropy so that the fuzz tests are run for 10min or so for every pull request, preventing buggy code from being merged.

I've opened a pull request to add a basic fuzz-testing harness here https://github.com/astropy/astropy/pull/14690. If you are keen on adding astropy to oss-fuzz I'd be happy to champion the integration :)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jesslatimer**
### Description

This pull request creates an adapter between hypothesis and atheris to make use of coverage-driven fuzzing.

EDIT: Fix #14691

### CODE REVIEW

**Comment by pllim** [comment]
Thanks! What is the motivation for this PR?

**Comment by pllim** [comment]
> The fuzzer will run forever

Probably not a good candidate for CI then. Perhaps can be some offline option people can run locally if they want.

**Comment by jesslatimer** [comment]
See this issue https://github.com/astropy/astropy/issues/14691 where I laid out a more detailed explanation of fuzz testing and the benefits of integrating with OSS-Fuzz.

>What is the motivation for this PR?

Atheris is a coverage-driven fuzzing engine. However, Hypothesis doesn't support coverage-driven feedback and so it has limited performance when it comes to automatically finding new code-paths (and therefore bugs).

_Black box fuzzing (Hypothesis)_
This approach to fuzzing is considered to be [open-loop](https://en.wikipedia.org/wiki/Control_system#Open-loop_and_closed-loop_control) and **does not** use feedback to improve future results.
```mermaid
flowchart LR
    A[Random Input] --> B[Fuzz Harness]
    B --> A
```
With black-box fuzzing the probability of finding a bug does not change over time.

_Grey box fuzzing (Atheris)_
This approach to fuzzing is considered to be [closed-loop](https://en.wikipedia.org/wiki/Control_system#Open-loop_and_closed-loop_control) and **does** use feedback to improve future results.
```mermaid
flowchart LR
    A[Random Input] --> B[Fuzz Harness]
    B --> C[Collect code coverage information]
    C -->  D[Mutate orgional input to maximize coverage]
    D --> B
```

With greybox fuzzing the probability of finding a bug over time increases.

>Probably not a good candidate for CI then.

I see CI integration working like this:
- This PR is edited and merged (although, I have read the contributing document and will need some guidance completing everything correctly)
- I open a PR with oss-fuzz for integration
- We use the [oss-fuzz github actions](https://google.github.io/oss-fuzz/getting-started/continuous-integration/)

Which will mean that:
- On new PR's the fuzzer will be run for 10min (on github actions) or so to catch shallow bugs before they make it into the repository.
- The fuzzer will be run on Google's distributed cluster (aka oss-fuzz) for a few hours a day in parallel, to catch deeper, hard-to-find bugs.

**Comment by mhvk** [comment]
I'm a bit skeptical that fuzz testing on astropy is going to be a net positive. The hypothesis tests found some bugs initially, but since then mostly found that rounding errors could be 1 bit more than previously foreseen, so that all that's really happened is that the tests had to be adjusted. I fear that extending this is just going to move us further in constantly having to judge whether nonsensical input is properly caught instead of focusing on finding actual physical problems or, really, actually implementing enhancements. Anyway, I guess that I would like to see a volunteer for dealing with test reports before we implement this.

File: astropy/time/tests/fuzz_precision.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    tp.test_timedelta_datetime_roundtrip.hypothesis.fuzz_one_input(
+        fdp.ConsumeBytes(byte_size))
+    tp.test_timedelta_from_parts.hypothesis.fuzz_one_input(
+        fdp.ConsumeBytes(byte_size))
+    tp.test_datetime_timedelta_sum.hypothesis.fuzz_one_input(
+        fdp.ConsumeBytes(byte_size))
+    tp.test_sidereal_lat_independent.hypothesis.fuzz_one_input(
+        fdp.ConsumeBytes(byte_size))
+    tp.test_sidereal_lon_independent.hypothesis.fuzz_one_input(
+       fdp.ConsumeBytes(byte_size))
```
[the remark this answers is not in the document]
Comment (jesslatimer) [reply astropy/time/tests/fuzz_precision.py]: I have split three of the tests into their own files in my latest push. Are they set up how you were asking?

File: astropy/time/tests/fuzz_precision.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+with atheris.instrument_imports():
+    import astropy.time.tests.test_precision as tp
+
+
+def TestOneInput(data):
+    fdp = atheris.FuzzedDataProvider(data)
+    byte_size = fdp.ConsumeIntInRange(0, 1000)
+    tp.test_abs_jd2_always_less_than_half_on_construction.hypothesis.fuzz_one_input(
+        fdp.ConsumeBytes(byte_size))
```
[the remark this answers is not in the document]
Comment (jesslatimer) [reply astropy/time/tests/fuzz_precision.py]: Done :)

**Comment by Zac-HD** [comment]
(unsubscribing, @ me if there's anything I can help with)

File: astropy/time/tests/test_precision.py [hunk 3]
```diff
@@ -696,14 +691,16 @@ def test_timedelta_conversion(scale1, scale2, jds_a, jds_b):
...
 
-@given(datetimes(), datetimes())  # datetimes have microsecond resolution
+@given(
+    scale=sampled_from(_utc_bad),
+    dt1=datetimes(),
+    dt2=datetimes()
+)  # datetimes have microsecond resolution
 @example(dt1=datetime(1235, 1, 1, 0, 0), dt2=datetime(9950, 1, 1, 0, 0, 0, 890773))
-@pytest.mark.parametrize("scale", _utc_bad)
 def test_datetime_difference_agrees_with_timedelta(scale, dt1, dt2):
```
[the remark this answers is not in the document]
Comment (jesslatimer) [reply astropy/time/tests/test_precision.py]: @Zac-HD I have potentially come up with a way around this issue while still using Atheris (see my latest push). I am almost certain my logic checks out. I have only done this for 3 of the tests so far.
For the `test_datetime_difference_agrees_with_timedelta` test, atheris appears to have found a bug that I could not track down.

I used some print statements to get some more information to show you an example.

The inputs were:
dt1:  datetime.datetime(1487, 12, 16, 9, 41, 41, 633307) 
dt2:  datetime.datetime(718, 1, 1, 0, 0, 0, 63766)
scale = utc

Which led to the `assert_almost_equal` function receiving:
- a (aka t2 - t1) =  -281221.4039533512
- b (aka TimeDelta(dt2 - dt1,...etc.)) =  -281222 days, 14:18:18.430459
- and calculated the difference between 'a' and 'b' to be 0.0 us

This is what doesn't make sense to me as there appears to be at least a day between 'a' and 'b', not zero.
This leads to no exception being raised by the test function, however, scale = UTC and an exception should always be raised in this case.

I'm happy to clarify anything and explore further if you need more information :)

**Pull request closed without merging**